core/chains/evm/logpoller/parser.go

```diff
@@ -0,0 +1,499 @@
+
+	where, err := v.whereClause(expressions, limiter)
+	if err != nil {
+		return "", nil, err
+	}
+
+	clauses = append(clauses, where)
+
+	order, err := v.orderClause(limiter)
+	if err != nil {
+		return "", nil, err
+	}
+
+	if len(order) > 0 {
+		clauses = append(clauses, order)
+	}
+
+	limit := v.limitClause(limiter)
+	if len(limit) > 0 {
+		clauses = append(clauses, limit)
+	}
+
+	return strings.Join(clauses, " "), v.args, nil
+}
+
+func (v *pgDSLParser) whereClause(expressions []query.Expression, limiter query.LimitAndSort) (string, error) {
+	segment := "WHERE evm_chain_id = :evm_chain_id"
+
+	if len(expressions) > 0 {
+		exp, err := v.combineExpressions(expressions, query.AND)
+		if err != nil {
+			return "", err
+		}
+
+		segment = fmt.Sprintf("%s AND %s", segment, exp)
+	}
+
+	if limiter.HasCursorLimit() {
+		var op string
+		switch limiter.Limit.CursorDirection {
+		case query.CursorFollowing:
+			op = ">"
+		case query.CursorPrevious:
+			op = "<"
+		default:
+			return "", errors.New("invalid cursor direction")
+		}
+
+		block, txHash, logIdx, err := valuesFromCursor(limiter.Limit.Cursor)
+		if err != nil {
+			return "", err
+		}
+
+		segment = fmt.Sprintf("%s AND block_number %s= :cursor_block AND tx_hash %s= :cursor_txhash AND log_index %s :cursor_log_index", segment, op, op, op)
+
+		v.args.withField("cursor_block_number", block).
+			withField("cursor_txhash", common.HexToHash(txHash)).
+			withField("cursor_log_index", logIdx)
+	}
+
+	return segment, nil
+}
+
+func (v *pgDSLParser) orderClause(limiter query.LimitAndSort) (string, error) {
+	sorting := limiter.SortBy
+
+	if limiter.HasCursorLimit() && !limiter.HasSequenceSort() {
+		var dir query.SortDirection
+
+		switch limiter.Limit.CursorDirection {
+		case query.CursorFollowing:
+			dir = query.Asc
+		case query.CursorPrevious:
+			dir = query.Desc
+		default:
+			return "", errors.New("unexpected cursor direction")
+		}
+
+		sorting = append(sorting, query.NewSortBySequence(dir))
+	}
+
+	if len(sorting) == 0 {
+		return "", nil
+	}
+
+	sort := make([]string, len(sorting))
+
+	for idx, sorted := range sorting {
+		var name string
+
+		order, err := orderToString(sorted.GetDirection())
+		if err != nil {
+			return "", err
+		}
+
+		switch sorted.(type) {
+		case query.SortByBlock:
+			name = blockFieldName
+		case query.SortBySequence:
+			sort[idx] = fmt.Sprintf("block_number %s, tx_hash %s, log_index %s", order, order, order)
+
+			continue
+		case query.SortByTimestamp:
+			name = timestampFieldName
+		default:
+			return "", errors.New("unexpected sort by")
+		}
+
+		sort[idx] = fmt.Sprintf("%s %s", name, order)
+	}
+
+	return fmt.Sprintf("ORDER BY %s", strings.Join(sort, ", ")), nil
+}
+
+func (v *pgDSLParser) limitClause(limiter query.LimitAndSort) string {
+	if !limiter.HasCursorLimit() && limiter.Limit.Count == 0 {
+		return ""
+	}
+
+	return fmt.Sprintf("LIMIT %d", limiter.Limit.Count)
+}
+
+func (v *pgDSLParser) getLastExpression() (string, error) {
```

Comment:
This is a bit odd to me. Get methods should be independent and not perform any cleanup/state reset in the background, right? Are you going to reuse that instance of `pgDSLParser` when building different queries? Probably new struct should be created every time you want to translate DSL to SQL so wiping out this state doesn't give any value, no?

Reply:
Huh, there is a huge overlap between this PR and https://github.com/smartcontractkit/chainlink/pull/12987, so I commented some items in the latter